File: options_lab/strategy/signal_backtest.py
"""Backtest an underlying-signal strategy that trades options.

Buy signal -> buy an ATM CALL. Sell signal -> buy an ATM PUT. Exit on the
opposite signal or at the session close, whichever comes first.

Four properties are structural rather than conventional, each because the
opposite has already produced a fake result somewhere in this project:

  NEXT-BAR ENTRY. A signal computed from bar i's close cannot be filled at that
  close - you only know it once the bar has ended. Entry is bar i+1's open. A
  signal on the last bar is dropped, not filled.

  ONE CONTRACT PER TRADE. The contract chosen at entry is the one marked and
  exited. Re-picking "the ATM" at exit would silently swap instruments and
  manufacture P&L out of the strike ladder.

  COSTS ALWAYS. net_pnl is computed from gross minus charges; there is no
  cost-free path. A flat option price must therefore produce a LOSS.

  NO INVENTED PRICES. If the required strike is not in the chain at that
  minute, the trade is skipped. It is never priced from a model.
"""
from __future__ import annotations

import numpy as np
import pandas as pd

from options_lab import costs

class AmbiguousChain(ValueError):
    """The chain carries more than one expiry.

    Strike selection used to read `chain['expiry'].iloc[0]`, so the traded
    expiry was whichever happened to sort first in the frame - chosen by
    accident, with no way for the caller to know which.
    """


TRADE_COLUMNS = [
    "entry_ts", "exit_ts", "direction", "right", "strike", "contract_id",
    "exit_contract_id", "entry_px", "exit_px", "underlying_entry",
    "underlying_exit", "gross_pnl", "cost", "net_pnl", "bars_held", "exit_reason",
]


def _empty() -> pd.DataFrame:
    return pd.DataFrame({c: pd.Series(dtype="object") for c in TRADE_COLUMNS})


def run(
    bars: pd.DataFrame,
    chain: pd.DataFrame,
    buys: np.ndarray,
    sells: np.ndarray,
    *,
    lot_size: int,
    lots: int = 1,
    regime: str = "quoted",
    max_strike_distance_pct: float = 0.02,
) -> pd.DataFrame:
    """One session. `bars` is the underlying at the signal timeframe."""
    if not (len(buys) == len(sells) == len(bars)):
        raise ValueError("signal arrays must align with bars")

    expiries = sorted(set(chain["expiry"])) if len(chain) else []
    if len(expiries) > 1:
        raise AmbiguousChain(
            f"chain carries {len(expiries)} expiries {expiries[:3]}; select one "
            f"before backtesting so the traded contract is chosen, not stumbled into"
        )

    # Price lookup: (contract_id, minute) -> close. Options print every minute,
    # so entries are snapped to the nearest available minute at or before entry.
    px = (chain.assign(ts=pd.to_datetime(chain["ts"]))
               .set_index(["contract_id", "ts"])["close"].sort_index())
    strikes = np.sort(chain["strike"].unique())

    trades: list[dict] = []
    open_pos: dict | None = None

    def quote(cid, ts):
        try:
            s = px.loc[cid]
        except KeyError:
            return None
        s = s[s.index <= ts]
        return float(s.iloc[-1]) if len(s) else None

    def close_out(pos, ts, spot, reason, i):
        exit_px = quote(pos["contract_id"], ts)
        if exit_px is None:
            return
        qty = lot_size * lots
        gross = (exit_px - pos["entry_px"]) * qty
        rt = costs.round_trip(premium=pos["entry_px"], lot_size=lot_size,
                              lots=lots, regime=regime)
        trades.append({
            "entry_ts": pos["entry_ts"], "exit_ts": ts,
            "direction": pos["direction"], "right": pos["right"],
            "strike": pos["strike"], "contract_id": pos["contract_id"],
            "exit_contract_id": pos["contract_id"],
            "entry_px": pos["entry_px"], "exit_px": exit_px,
            "underlying_entry": pos["spot"], "underlying_exit": spot,
            "gross_pnl": gross, "cost": rt.total, "net_pnl": gross - rt.total,
            "bars_held": i - pos["bar"], "exit_reason": reason,
        })

    n = len(bars)
    for i in range(n):
        ts = bars.index[i]
        spot = float(bars["Close"].iloc[i])

        # Exit first: an opposite signal on bar i-1 closes at bar i's open.
        if open_pos is not None and i > 0:
            opp = sells[i - 1] if open_pos["direction"] == "long_call" else buys[i - 1]
            if opp:
                close_out(open_pos, ts, spot, "opposite_signal", i)
                open_pos = None

        if i == 0 or open_pos is not None:
            continue

        want_call, want_put = bool(buys[i - 1]), bool(sells[i - 1])
        if not (want_call or want_put):
            continue

        right = "CE" if want_call else "PE"
        entry_spot = float(bars["Open"].iloc[i])
        k = float(strikes[np.abs(strikes - entry_spot).argmin()])
        if abs(k - entry_spot) / entry_spot > max_strike_distance_pct:
            continue

        cid = f"{chain['contract_id'].iloc[0].split('|')[0]}|" \
              f"{chain['expiry'].iloc[0]:%Y-%m-%d}|{k:g}|{right}"
        entry_px = quote(cid, ts)
        if entry_px is None or entry_px <= 0:
            continue

        open_pos = {
            "entry_ts": ts, "direction": "long_call" if want_call else "long_put",
            "right": right, "strike": k, "contract_id": cid,
            "entry_px": entry_px, "spot": entry_spot, "bar": i,
        }

    if open_pos is not None:
        close_out(open_pos, bars.index[-1], float(bars["Close"].iloc[-1]),
                  "session_close", n - 1)

    return pd.DataFrame(trades, columns=TRADE_COLUMNS) if trades else _empty()

File: android/engine/src/main/kotlin/com/optionslab/engine/SignalBacktest.kt
package com.optionslab.engine

import kotlin.math.abs

/**
 * Backtest an underlying-signal strategy that trades options. Port of
 * `strategy/signal_backtest.py`.
 *
 * Buy signal -> buy an ATM CALL. Sell signal -> buy an ATM PUT. Exit on the
 * opposite signal or at the session close. NEXT-BAR ENTRY, ONE CONTRACT PER
 * TRADE, COSTS ALWAYS, NO INVENTED PRICES.
 */
object SignalBacktest {
    class AmbiguousChain(msg: String) : IllegalArgumentException(msg)

    data class Bars(val minutes: IntArray, val open: DoubleArray, val high: DoubleArray, val low: DoubleArray, val close: DoubleArray)

    data class Trade(
        val entryMinute: Int, val exitMinute: Int, val direction: String, val right: Right, val strike: Double,
        val contractId: String, val entryPx: Double, val exitPx: Double,
        val underlyingEntry: Double, val underlyingExit: Double,
        val grossPnl: Double, val cost: Double, val netPnl: Double, val barsHeld: Int, val exitReason: String,
    )

    fun run(
        bars: Bars, chain: List<Series>, buys: BooleanArray, sells: BooleanArray,
        lotSize: Int, lots: Int = 1, regime: String = "quoted", maxStrikeDistancePct: Double = 0.02,
        underlying: String = "NIFTY",
    ): List<Trade> {
        require(buys.size == sells.size && sells.size == bars.close.size) { "signal arrays must align with bars" }
        val opts = chain.filter { it.right != Right.IX }
        val expiries = opts.mapNotNull { it.expiry }.toSortedSet()
        if (expiries.size > 1) throw AmbiguousChain(
            "chain carries ${expiries.size} expiries ${expiries.take(3)}; select one before backtesting so the traded contract is chosen, not stumbled into")
        val byKey = opts.associateBy { it.strike to it.right }
        val strikes = opts.map { it.strike }.distinct().sorted()
        val trades = ArrayList<Trade>()

        fun quote(s: Series, minute: Int): Double? {
            val i = s.lastAtOrBefore(minute)
            return if (i >= 0) s.close[i] else null
        }

        data class Pos(val entry: Int, val direction: String, val right: Right, val strike: Double, val series: Series, val px: Double, val spot: Double, val bar: Int)

        fun closeOut(p: Pos, minute: Int, spot: Double, reason: String, i: Int) {
            val exitPx = quote(p.series, minute) ?: return
            val qty = lotSize * lots
            val gross = (exitPx - p.px) * qty
            val rt = Costs.roundTrip(p.px, lotSize, lots, regime)
            trades += Trade(p.entry, minute, p.direction, p.right, p.strike, p.series.contractId(underlying),
                p.px, exitPx, p.spot, spot, gross, rt.total, gross - rt.total, i - p.bar, reason)
        }

        var open: Pos? = null
        val n = bars.close.size
        for (i in 0 until n) {
            val ts = bars.minutes[i]
            val spot = bars.close[i]
            if (open != null && i > 0) {
                val opp = if (open.direction == "long_call") sells[i - 1] else buys[i - 1]
                if (opp) { closeOut(open, ts, spot, "opposite_signal", i); open = null }
            }
            if (i == 0 || open != null) continue
            val wantCall = buys[i - 1]
            val wantPut = sells[i - 1]
            if (!(wantCall || wantPut) || strikes.isEmpty()) continue
            val right = if (wantCall) Right.CE else Right.PE
            val entrySpot = bars.open[i]
            val k = strikes.minBy { abs(it - entrySpot) }
            if (abs(k - entrySpot) / entrySpot > maxStrikeDistancePct) continue
            val s = byKey[k to right] ?: continue
            val px = quote(s, ts) ?: continue
            if (px <= 0) continue
            open = Pos(ts, if (wantCall) "long_call" else "long_put", right, k, s, px, entrySpot, i)
        }
        open?.let { closeOut(it, bars.minutes[n - 1], bars.close[n - 1], "session_close", n - 1) }
        return trades
    }
}

/**
 * Signed option order flow (port of `features/flow.py`): per minute, the sum
 * over the chain of sign(close_t - close_{t-1}) * volume, calls +1, puts -1.
 * An unchanged price contributes zero however large the volume.
 */
object Flow {
    fun signedFlow(chain: List<Series>, normalise: Boolean = false): Map<Int, Double> {
        val flow = java.util.TreeMap<Int, Double>()
        val traded = java.util.TreeMap<Int, Double>()
        for (s in chain) {
            val side = when (s.right) { Right.CE -> 1.0; Right.PE -> -1.0; Right.IX -> continue }
            val vol = s.volume ?: continue
            for (i in 0 until s.size) {
                val tick = if (i == 0) 0.0 else Math.signum(s.close[i] - s.close[i - 1])
                val m = s.minutes[i]
                flow.merge(m, tick * side * vol[i], Double::plus)
                traded.merge(m, vol[i].toDouble(), Double::plus)
            }
        }
        if (!normalise) return flow
        return flow.mapValues { (m, f) -> val t = traded[m] ?: 0.0; if (t == 0.0) 0.0 else f / t }
    }
}

/** Regime gates (port of `features/gates.py`): where a signal may trade, not what it says. */
object Gates {
    const val SESSION_MINUTES = 375
    const val LIQUID_THRESHOLD = 0.80
    class Expired(msg: String) : IllegalArgumentException(msg)

    fun dte(session: java.time.LocalDate, expiry: java.time.LocalDate): Int {
        val d = (expiry.toEpochDay() - session.toEpochDay()).toInt()
        if (d < 0) throw Expired("contract expired $expiry, session is $session")
        return d
    }

    fun dteBucket(days: Int): String {
        if (days < 0) throw Expired("dte $days is negative; the contract already expired")
        for ((edge, label) in listOf(0 to "0", 1 to "1", 3 to "2-3", 7 to "4-7", 21 to "8-21")) if (days <= edge) return label
        return "22+"
    }

    fun atmThetaSharePerDay(dte: Int): Double {
        require(dte > 0) { "theta share is undefined on expiry day - T is 0 and the option settles rather than decays." }
        val t = dte / 365.0
        return (1.0 / (2.0 * t)) / 365.0
    }

    fun spreadInThetaMinutes(spreadPts: Double, premium: Double, dte: Int): Double {
        val decay = atmThetaSharePerDay(dte) * premium
        require(decay > 0) { "non-positive decay" }
        return spreadPts / (decay / SESSION_MINUTES)
    }

    fun timeOfDayBucket(minute: Int): String = when {
        minute < 10 * 60 -> "open"
        minute < 14 * 60 + 30 -> "midday"
        else -> "close"
    }

    fun liquid(tradedFraction: Double, threshold: Double = LIQUID_THRESHOLD): Boolean {
        require(!tradedFraction.isNaN()) { "traded_fraction is NaN; compute it before gating" }
        require(tradedFraction in 0.0..1.0) { "traded_fraction $tradedFraction is outside [0, 1]" }
        return tradedFraction >= threshold
    }
}
